File: OmniVoice-Studio/frontend/src/utils/ssmlLite.js
/**
 * SSML-LITE (client port) of backend/services/ssml_lite.py.
 *
 * Parity is no longer kept by hand: the canonical longform grammar (#27) is
 * exercised end-to-end by the shared golden corpus in
 * tests/fixtures/longform_parser_cases.json, asserted byte-for-byte against
 * BOTH this port (via longformParser.js → storyToSpans) and the Python parser.
 * A drift between the two SSML impls fails one of those two suites.
 *
 * Splits one narration line into ordered prosody segments so the Stories Editor
 * compiles the same `[slow]/[fast]/[emphasis]/[spell]` markup the Audiobook
 * backend parser honours. Pure; no deps.
 *
 *   parseSsmlLite(text) -> [{ text, speed, spell, emphasis }, …]
 *
 * Plain text → one segment {speed:null, spell:false, emphasis:false}. Tags nest
 * (innermost wins per property); an unclosed tag runs to end-of-line; a stray
 * close is ignored; adjacent identical-prosody segments merge.
 */
export const SLOW_SPEED = 0.85;
export const FAST_SPEED = 1.15;
const EMPHASIS_SPEED = 0.92;

const TAGS = {
  slow: { speed: SLOW_SPEED, spell: null, emphasis: null },
  fast: { speed: FAST_SPEED, spell: null, emphasis: null },
  emphasis: { speed: EMPHASIS_SPEED, spell: null, emphasis: true },
  spell: { speed: null, spell: true, emphasis: null },
};

// Fixed-literal alternation, no quantifier overlap → linear-time (ReDoS-safe).
const TAG_RE = /\[(\/?)(slow|fast|emphasis|spell)\]/gi;

function resolve(stack) {
  let speed = null;
  let spell = false;
  let emphasis = false;
  for (const name of stack) {
    const spec = TAGS[name];
    if (spec.speed !== null) speed = spec.speed;
    if (spec.spell !== null) spell = !!spec.spell;
    if (spec.emphasis !== null) emphasis = !!spec.emphasis;
  }
  return { speed, spell, emphasis };
}

export function parseSsmlLite(text) {
  if (!text) return [];
  if (!text.includes('[')) return [{ text, speed: null, spell: false, emphasis: false }];

  const segments = [];
  const stack = [];
  let last = 0;

  const emit = (chunk) => {
    if (!chunk) return;
    const props = resolve(stack);
    const prev = segments[segments.length - 1];
    if (
      prev &&
      prev.speed === props.speed &&
      prev.spell === props.spell &&
      prev.emphasis === props.emphasis
    ) {
      prev.text += chunk;
      return;
    }
    segments.push({ text: chunk, ...props });
  };

  const re = new RegExp(TAG_RE.source, TAG_RE.flags);
  let m;
  while ((m = re.exec(text)) !== null) {
    emit(text.slice(last, m.index));
    last = m.index + m[0].length;
    const isClose = m[1] === '/';
    const name = m[2].toLowerCase();
    if (isClose) {
      for (let i = stack.length - 1; i >= 0; i--) {
        if (stack[i] === name) {
          stack.splice(i, 1);
          break;
        }
      }
    } else {
      stack.push(name);
    }
  }
  emit(text.slice(last));
  return segments;
}

/** Space out a run for [spell]: "USA" → "U S A". */
export function spellOut(word) {
  return (word || '').split(/\s+/).join('').split('').join(' ');
}

File: OmniVoice-Studio/backend/services/ssml_lite.py
"""SSML-LITE — inline prosody/spell markup for longform lines (PR 8).

A *single line* of narration may carry inline tags that nudge the engine's
delivery without reaching for full SSML:

  * ``[slow]…[/slow]``         — speak slower   (``speed ≈ 0.85``)
  * ``[fast]…[/fast]``         — speak faster   (``speed ≈ 1.15``)
  * ``[emphasis]…[/emphasis]`` — mild emphasis: a gentle slow-down
                                 (``speed ≈ 0.92``) plus an ``emphasis`` flag
                                 the caller may use for future markup
  * ``[spell]…[/spell]``       — spell the run out letter-by-letter
                                 (``spell=True``; the caller spaces the chars)

:func:`parse_ssml_lite` splits one line into ordered segments::

    [{"text": str, "speed": float | None, "spell": bool, "emphasis": bool}, …]

Semantics (kept deliberately small and predictable):

  * Plain text → exactly one segment ``{text, speed=None, spell=False,
    emphasis=False}``.
  * Tags nest; the **innermost** tag wins for any property it sets. ``speed``
    from an inner ``[fast]`` overrides an outer ``[slow]``; ``[spell]`` inside
    ``[slow]`` keeps the slow speed *and* turns spelling on.
  * An **unclosed** tag applies to the end of the line.
  * A stray close tag with no matching open is ignored (treated as literal
    nothing — the markers are always stripped from the emitted ``text``).
  * Adjacent segments that share identical (speed, spell, emphasis) are merged
    so plain runs stay single segments.

This module is pure (no torch, no I/O) so it is cheap to import and unit-test.
The regex is ReDoS-safe: it is a fixed alternation of literal tag tokens with
no quantifier overlap, so matching is linear in the input length.
"""

from __future__ import annotations

import re
from typing import Optional

# Speed multipliers. ``None`` means "engine default" (no override emitted).
SLOW_SPEED = 0.85
FAST_SPEED = 1.15
# Emphasis maps to a *mild* slow-down (between default and [slow]) plus a flag.
EMPHASIS_SPEED = 0.92

# Recognised tag names → the (speed_delta, spell, emphasis) they impose while
# open. ``speed`` of ``None`` for a tag means "this tag does not touch speed".
_TAGS: dict[str, dict] = {
    "slow": {"speed": SLOW_SPEED, "spell": None, "emphasis": None},
    "fast": {"speed": FAST_SPEED, "spell": None, "emphasis": None},
    "emphasis": {"speed": EMPHASIS_SPEED, "spell": None, "emphasis": True},
    "spell": {"speed": None, "spell": True, "emphasis": None},
}

# One regex that matches any open/close tag for the known names. It is a plain
# alternation of fixed literals — ``\[/?(?:slow|fast|emphasis|spell)\]`` — with
# no nested quantifiers and no overlapping ``*``/``+`` runs, so it cannot
# backtrack polynomially (ReDoS-safe). ``finditer`` walks it left-to-right.
_TAG_RE = re.compile(
    r"\[(/?)(" + "|".join(re.escape(name) for name in _TAGS) + r")\]",
    re.IGNORECASE,
)


def _resolve(stack: list[str]) -> dict:
    """Collapse an open-tag stack into the effective segment properties.

    Outer→inner walk: a later (more-deeply-nested) tag overrides any property
    it sets, leaving untouched properties from outer tags intact. So
    ``[slow][spell]`` yields ``speed=SLOW_SPEED, spell=True``.
    """
    speed: Optional[float] = None
    spell = False
    emphasis = False
    for name in stack:
        spec = _TAGS[name]
        if spec["speed"] is not None:
            speed = spec["speed"]
        if spec["spell"] is not None:
            spell = bool(spec["spell"])
        if spec["emphasis"] is not None:
            emphasis = bool(spec["emphasis"])
    return {"speed": speed, "spell": spell, "emphasis": emphasis}


def parse_ssml_lite(text: str) -> list[dict]:
    """Split one line of SSML-LITE markup into ordered prosody segments.

    See the module docstring for the full contract. Always returns at least one
    segment for non-empty input; returns ``[]`` for ``None``/empty input.
    """
    if not text:
        return []
    if "[" not in text:
        return [{"text": text, "speed": None, "spell": False, "emphasis": False}]

    segments: list[dict] = []
    stack: list[str] = []
    last = 0

    def emit(chunk: str) -> None:
        if not chunk:
            return
        props = _resolve(stack)
        seg = {"text": chunk, **props}
        # Merge with the previous segment when prosody is identical so plain
        # text never fragments into multiple identical-styled pieces.
        if segments:
            prev = segments[-1]
            if (
                prev["speed"] == seg["speed"]
                and prev["spell"] == seg["spell"]
                and prev["emphasis"] == seg["emphasis"]
            ):
                prev["text"] += chunk
                return
        segments.append(seg)

    for m in _TAG_RE.finditer(text):
        emit(text[last:m.start()])
        last = m.end()
        is_close = m.group(1) == "/"
        name = m.group(2).lower()
        if is_close:
            # Close the nearest matching open tag; ignore an unmatched close.
            for i in range(len(stack) - 1, -1, -1):
                if stack[i] == name:
                    del stack[i]
                    break
        else:
            stack.append(name)  # unclosed opens stay on the stack to EOL

    emit(text[last:])

    if not segments:
        # Input was only tag markers (e.g. "[slow][/slow]"): nothing to speak.
        return []
    return segments


def spell_out(word: str) -> str:
    """Space out a run for the ``[spell]`` case: ``"USA"`` → ``"U S A"``.

    Collapses surrounding whitespace, then joins the remaining characters with
    single spaces so the engine pronounces each letter discretely. Whitespace
    inside the run is treated as a separator (each token spelled, joined by a
    single space), so ``"go USA"`` → ``"g o U S A"``.
    """
    if not word:
        return ""
    # Drop all existing whitespace, then interleave the visible characters with
    # spaces. ``split()`` + ``"".join`` removes runs of whitespace first.
    compact = "".join(word.split())
    return " ".join(compact)
